frontend/src/lib/utils.ts

```diff
@@ -203,3 +203,93 @@ export function labelVisibilityProtoValueToEnum(
       return LabelVisibilityOptions.Visible
   }
 }
+
+/**
+ * Looks for an IFrame with given className inside given querySet
+ */
+export function findAnIFrameWithClassName(
+  qs: NodeListOf<HTMLIFrameElement> | HTMLCollectionOf<HTMLIFrameElement>,
+  className: string
+): HTMLIFrameElement | null {
+  for (let i = 0; i < qs.length; i++) {
+    const cd = qs[i].contentDocument
+    if (cd && cd.getElementsByClassName(className).length > 0) {
+      return qs[i]
+    }
+  }
+  return null
+}
+
+/**
+ * Returns True if IFrame can be accessed otherwise returns False
+ */
+export function canAccessIFrame(iframe: HTMLIFrameElement): boolean {
+  try {
+    if (iframe.contentWindow === null) return false
+    const doc = iframe.contentDocument || iframe.contentWindow.document
+    const html = doc.body.innerHTML
+    return html !== null && html !== ""
+  } catch (err) {
+    return false
+  }
+}
+
+/**
+ * Tries to get an IFrame in which Streamlit app is embedded on Cloud deployments.
+ * It assumes iframe has title="streamlitApp", iterates over IFrames,
+ * and looks which IFrame contains div with stAppId value, otherwise returns first found iFrame or null.
+ */
+export function getIFrameEnclosingApp(
+  embeddingId: string
+): HTMLIFrameElement | null {
+  if (!isInChildFrame()) {
+    return null
+  }
+  const embeddingIdClassName = getEmbeddingIdClassName(embeddingId)
+  const qsStreamlitAppStr = 'iframe[title="streamlitApp"]'
```

Comment:
If this query string is changed to be just `'iframe'`, would that be equivalent to calling `window.document.getElementsByTagName("iframe")`? If so, then I think this can be generalized more so that this case and the code from L263 and below can be handled by a single helper function

Reply:
I got your point, but I have done it on purpose `querySelectorAll` is called first, and the first case is very specific, it looks for `iframe[title="streamlitApp"]` and looks for `embeddedId` then, I've checked it on Streamlit Cloud and the app always hits this path of exection.

Later we have fallbacks, and I really want to use `getElementsByTagName` instead of `querySelectorAll`, it's a different method and has a different result (1st returns `NodeListOf<HTMLIFrameElement>`, 2nd returns `HTMLCollectionOf<HTMLIFrameElement>`). It's not strict it just looks for any first found IFrame, so I don't want to iterate and look for `embeddedId` on purpose. These lines were never hit, unless I've set `embeddedId` to constant on purpose, but they are much broader than first look for an IFrame.

I know it looks messy, but it was done specifically to avoid issues with access to HTMLElements and to find an IFrame despite the execution context

Reply:
I want to use different method for querying 'iframe[title="streamlitApp"]' and for looking for `iframe`, 1st uses `querySelectorAll` which returns `NodeListOf<HTMLIFrameElement>`, 2nd uses `getElementsByTagName` which returns `HTMLCollectionOf<HTMLIFrameElement>`. 

I guess I could write 1 method, which will have an conditional inside, to use different methods based on arguments, but I am not sure if it's going to be more readable